beam/beam_job_service.sh

```diff
@@ -0,0 +1,140 @@
+  /usr/share/google/get_metadata_value "attributes/${ARTIFACTS_DIR_METADATA_KEY}" \
+    || echo "gs://$(/usr/share/google/get_metadata_value "attributes/dataproc-bucket")/beam-artifacts"
+}
+
+function download_snapshot() {
+  readonly snapshot_url="${1}"
+  readonly protocol="$(echo "${snapshot_url}" | head -c5)"
+  if [ "${protocol}" = "gs://" ]; then
+    gsutil cp "${snapshot_url}" "${LOCAL_JAR_NAME}"
+  else
+    curl -o "${LOCAL_JAR_NAME}" "${snapshot_url}"
+  fi
+}
+
+function flink_master_url() {
+  local start_flink_yarn_session="$(/usr/share/google/get_metadata_value \
+    "attributes/${START_FLINK_YARN_SESSION_METADATA_KEY}" \
+    || echo "${START_FLINK_YARN_SESSION_DEFAULT}")"
+  # TODO: delete this workaround when the beam job service is able to understand 
+  # flink in yarn mode.
+  if ${start_flink_yarn_session} ; then
+    # grab final field from the first yarn application that contains 'flink'
+    yarn application -list >&2 # Logging this here is useful for debugging potential failures
+    yarn application -list \
+      | grep -i 'flink' \
+      | head -n1 \
+      | awk -F $'\t' '{print $9}' \
+      | cut -c8-
+  else
+    echo "localhost:8081"
+  fi
+}
+
+function install_job_service() {
+  local master_url="$(/usr/share/google/get_metadata_value attributes/dataproc-master)"
+  local artifacts_dir="$(get_artifacts_dir)"
+  local release_snapshot_url="$(/usr/share/google/get_metadata_value \
+    "attributes/${RELEASE_SNAPSHOT_URL_METADATA_KEY}" \
+    || echo "${RELEASE_SNAPSHOT_URL_DEFAULT}")"
+
+  echo "Retrieving Beam Job Service snapshot from ${release_snapshot_url}"
+
+  local flink_master="$(flink_master_url)"
+  echo "resolved flink master to: '${master_url}'"
+
+  mkdir -p "${SERVICE_INSTALL_DIR}"
+  pushd "${SERVICE_INSTALL_DIR}"
+  download_snapshot "${release_snapshot_url}"
+  popd
+  mkdir -p "${SERVICE_WORKING_DIR}"
+  chown -R "${SERVICE_WORKING_USER}" "${SERVICE_WORKING_DIR}"
+
+  cat > "/etc/systemd/system/beam-job-service.service" <<EOF
+[Unit]
+Description=Beam Job Service
+After=default.target
+
+[Service]
+Type=simple
+User=${SERVICE_WORKING_USER}
+WorkingDirectory=${SERVICE_WORKING_DIR}
+ExecStart=/usr/bin/java \
+  -jar ${SERVICE_INSTALL_DIR}/${LOCAL_JAR_NAME} \
+  --job-host=${master_url}\
+  --artifacts-dir=${artifacts_dir} \
+  --flink-master-url=${flink_master}
+Restart=on-failure
+
+[Install]
+WantedBy=multi-user.target
+EOF
+systemctl enable beam-job-service
+}
+
+function run_job_service() {
+  systemctl restart beam-job-service
+}
+
+function pull_beam_images() {
+  local beam_image_version="$(/usr/share/google/get_metadata_value \
+    "attributes/${BEAM_IMAGE_VERSION_METADATA_KEY}" \
+    || echo "${BEAM_IMAGE_VERSION_DEFAULT}")"
+  local image_repo="$(/usr/share/google/get_metadata_value \
+    "attributes/${BEAM_IMAGE_REPOSITORY_KEY}" \
+    || echo "${BEAM_IMAGE_REPOSITORY_DEFAULT}")"
+  sudo -u yarn -i docker pull "${image_repo}/go:${beam_image_version}"
```

Comment:
Please add a comment as to why the `-i` flag is necessary (presumably to set `HOME` etc. correctly). In most cases, this would _not_ be desired due to inheritance of `.profile` and other configuration files.

Reply:
Ack. will do.  TL;DR: GCR authorization is weird, and happens on a per-user basis.